modules/core/src/main/scala/Versioning.scala

```diff
@@ -0,0 +1,276 @@
+package schemaz
+
+trait Versioning[R <: Realisation] extends SchemaModule[R] {
+
+  final val Current: Version[Unit, Unit] = new Version(())
+
+  sealed case class Version[Types, Re](registry: Re)(
+    implicit build: Version.Build[Types, Re]
+  ) {
+
+    lazy val types: Types = build(registry)
+
+    def schema[A, RA](leaf: SchemaZ[RA, A])(
+      implicit add: Version.AddEntry[Types, Unit, A, RA]
+    ): Version[(Schema[A], Types), (Version.Entry.Aux[A, RA, Types], Re)] =
+      new Version((add(Version.Entry((_: Unit) => leaf)), registry))
+
+    def schema[A, RA, D](ctr: D => SchemaZ[RA, A])(
+      implicit add: Version.AddEntry[Types, D, A, RA]
+    ): Version[(Schema[A], Types), (Version.Entry.Aux[A, RA, Types], Re)] =
+      new Version(
+        (add(Version.Entry(ctr)), registry)
+      )
+
+    def schema[A, RA, D1, D2](ctr: (D1, D2) => SchemaZ[RA, A])(
+      implicit add: Version.AddEntry[Types, (D1, D2), A, RA]
+    ): Version[(Schema[A], Types), (Version.Entry.Aux[A, RA, Types], Re)] =
+      new Version(
+        (add(Version.Entry(ctr.tupled)), registry)
+      )
+
+    def schema[A, RA, D1, D2, D3](ctr: (D1, D2, D3) => SchemaZ[RA, A])(
+      implicit add: Version.AddEntry[Types, (D1, D2, D3), A, RA]
+    ): Version[(Schema[A], Types), (Version.Entry.Aux[A, RA, Types], Re)] =
+      new Version(
+        (add(Version.Entry(ctr.tupled)), registry)
+      )
+
+    def schema[A, RA, D1, D2, D3, D4](ctr: (D1, D2, D3, D4) => SchemaZ[RA, A])(
+      implicit add: Version.AddEntry[Types, (D1, D2, D3, D4), A, RA]
+    ): Version[(Schema[A], Types), (Version.Entry.Aux[A, RA, Types], Re)] =
+      new Version(
+        (add(Version.Entry(ctr.tupled)), registry)
+      )
+
+    def schema[A, RA, D1, D2, D3, D4, D5](ctr: (D1, D2, D3, D4, D5) => SchemaZ[RA, A])(
```

Comment:
Code gen to do arity > 5? Only has to be written once...

Reply:
Once we are confident this approach will solve the problem correctly and completely, we can generate overloads for arities > 5.